### ISSUE TITLE
Release 2.0

### ISSUE THREAD

**Issue by MusicDin** [terraform-lxd/terraform-provider-lxd#396]
This issue is intended as a discussion for 2.0 release.

---

We should document somewhere attributes that have been changed/removed across the resources. Shell we include that in a release note or as a small section in provider's documentation?

Renamed attributes:
- `lxd_provider.lxd_remote` -> `remote`
- `lxd_instance_file.target_file` -> `target_path`
- `lxd_instance_file.source` -> `source_path`
- `lxd_instance.start_on_create` -> `running`
- `lxd_instance.file.target_file` -> `target_path`
- `lxd_instance.file.source` -> `source_path`
- `lxd_instance.file.instance_name` -> `instance`
- `lxd_publish_image.container` -> `instance`
- `lxd_snapshot.container_name` -> `instance`

Removed attributes:
- `lxd_provider.refresh_interval`
- All attributes that were already marked as deprecated in 1.x.
  - `lxd_instance.privileged` 
  - `lxd_project.target` 
  - `lxd_snapshot.creation_date`
  - `lxd_publish_image.container`

Removed deprecated resources:
- `lxd_container`
- `lxd_container_file`
- `lxd_volume_container_attach`

---

These should be part of the 2.0:
- [x] https://github.com/terraform-lxd/terraform-provider-lxd/pull/386
- [x] https://github.com/terraform-lxd/terraform-provider-lxd/pull/389

Should we also update [known limitations](https://github.com/terraform-lxd/terraform-provider-lxd#known-limitations) in readme to include exec as an option for running commands within an instance?

---

Please update the list if I have missed something.

**Comment by matifali** [comment]
Is there any ETA for when 2.0 will be released?

> Should we also update [known limitations](https://github.com/terraform-lxd/terraform-provider-lxd#known-limitations) in the readme to include exec as an option for running commands within an instance?

I would request adding this as part of 2.0 as it will allow running custom commands when changing the state of an existing container, for example, on each start of a stopped instance. `cloud-init.user_data` only runs at first boot when the instance is created.

```hcl
resource "lxd_instance" "demo" {
	running = "true|false"
	...
}
```

**Comment by MusicDin** [comment]
Beside some small things that have to be done, I think `2.0` should be released within next 2 weeks.

Regarding `exec`, it is currently implemented, but the command is only invoked if the command itself is changed (differs compared to terraform state) or if `triggers` attribute is changed (where triggers is a list of arbitrary strings).

I think that with our current approach, it will not be possible to run the command on each instance start. So, this should be reviewed an potentially improved before the release to avoid immediate breaking changes to the exec just after the release.

Instead of triggers, we could use attributes like `run` and `run_if`, where `run` would have some predetermined options [*once*, *on_start*, *always*, ...] and `run_if` would be an optional constraint that would allow disabling the command execution if set to false (more like *enabled*/*disabled*). Any suggestions on this topic would be very welcome.

Here are the unreleased docs for exec: https://github.com/terraform-lxd/terraform-provider-lxd/blob/main/docs/resources/instance.md#executing-commands-in-instances

**Comment by matifali** [comment]
@MusicDin The context behind this request is that we want to run the coder_agent.init_script on each start. The script is usually static unless the agent gets updated, but the script uses `CODER_AGENT_TOKEN` that is refreshed on each start. I am looking for something equivalent to Docker's `entrypoint` that can execute dynamic script at each start of the instance. The script is provided by a terraform resource `coder_agent` and will passed as 
`coder_agent.example.init_script`

So to summarise, we need 2 things,
1. the ability to run a script as `entrypoint`.
2. Update the value of an environment variable. This can be hacked with something like `export CODER_AGENT_TOKEN=${coder_agent.example.token} && sh -c ${coder_agent.example.init_script}`

Feel free to let me know if you need more context. I can share a working example template that recreates the container instead of start/stop.

**Comment by MusicDin** [comment]
If script is more then just one command, it may be more feasible to upload the script to the instance and just start it using exec. Something like this:
```hcl
resource "lxd_instance" "workspace" {
  name  = "coder-workspace"
  image = "ubuntu:22.04"

  file {
    content     = coder_agent.example.init_script
    target_path = "/init-script.sh"
    mode        = "0700"
  }

  exec {
    command = ["/bin/sh", "/init-script.sh"]

    environment = {
      "CODER_AGENT_TOKEN" = coder_agent.example.token
    }
  }
}
```

Also, if the instance can be restarted outside the scope of terraform, you will need to configure the script to be started on instance boot (independent of terraform).

**Comment by matifali** [comment]
@MusicDin Everything is managed via terraform. Thanks for the tip. I am looking forward to seeing the `exec` functionality.

**Comment by MusicDin** [comment]
Yes, it should be this week.

**Comment by MusicDin** [comment]
2.0.0 is released now.
